File: src/python-bindings/negotiator.cpp

// Note - python_bindings_common.h must be included before condor_common to avoid
// re-definition warnings.
#include "python_bindings_common.h"

#include "condor_adtypes.h"
#include "dc_collector.h"
#include "condor_version.h"

#include <memory>

#include "old_boost.h"
#include "classad_wrapper.h"
#include "module_lock.h"
#include "htcondor.h"
#include "daemon_location.h"

using namespace boost::python;

static boost::python::list
toList(const boost::shared_ptr<classad::ClassAd> ad, const std::vector<std::string> &attrs)
{
    int idx = 1;
    bool hasattr = true;
    boost::python::list result;
    while (hasattr)
    {
        hasattr = false;
        boost::shared_ptr<ClassAdWrapper> nextAd(new ClassAdWrapper());
        for (std::vector<std::string>::const_iterator it = attrs.begin(); it != attrs.end(); it++)
        {
            std::stringstream attr; attr << *it << idx;
            classad::ExprTree *expr = NULL;
            if ((expr = ad->Lookup(attr.str())))
            {
                classad::ExprTree *copy = expr->Copy();
                if (!copy) { THROW_EX(HTCondorInternalError, "Unable to create copy of ClassAd expression"); }
                if (!nextAd->Insert(*it, copy)) { THROW_EX(HTCondorInternalError, "Unable to copy attribute into destination ClassAd"); }
                hasattr = true;
            }
        }
        if (hasattr)
        {
            result.append(nextAd);
        }
        idx++;
    }
    return result;
}


struct Negotiator {

    void use_local_negotiator()
    {
        Daemon neg( DT_NEGOTIATOR, 0, 0 );
        bool result;
        {
        condor::ModuleLock ml;
        result = neg.locate();
        }

        if (result)
        {
            if (neg.addr())
            {
                m_addr = neg.addr();
            }
            else
            {
                THROW_EX(HTCondorLocateError, "Unable to locate negotiator address.");
            }
            m_version = neg.version() ? neg.version() : "";
        }
        else
        {
            THROW_EX(HTCondorLocateError, "Unable to locate local daemon");
        }
    }

    Negotiator() {
        use_local_negotiator();
    }

    Negotiator(boost::python::object loc)
      : m_addr(), m_version("")
    {
		int rv = construct_for_location(loc, DT_NEGOTIATOR, m_addr, m_version);
		if (rv == 0) {
			use_local_negotiator();
		} else if (rv < 0) {
			if (rv == -2) { boost::python::throw_error_already_set(); }
			THROW_EX(HTCondorValueError, "Unknown type");
		}
	}

    ~Negotiator()
    {
    }

    boost::python::object location() const {
        return make_daemon_location(DT_NEGOTIATOR, m_addr, m_version);
    }

    void setPriority(const std::string &user, float prio)
    {
        if (prio < 0) THROW_EX(HTCondorValueError, "User priority must be non-negative");
        sendUserValue(SET_PRIORITY, user, prio);
    }

    void setFactor(const std::string &user, float factor)
    {
        if (factor<1) THROW_EX(HTCondorValueError, "Priority factors must be >= 1");
        sendUserValue(SET_PRIORITYFACTOR, user, factor);
    }

    void setUsage(const std::string &user, float usage)
    {
        if (usage < 0) THROW_EX(HTCondorValueError, "Usage must be non-negative.");
        sendUserValue(SET_ACCUMUSAGE, user, usage);
    }

    void setBeginUsage(const std::string &user, time_t time)
    {
        sendUserValue(SET_BEGINTIME, user, time);
    }

    void setLastUsage(const std::string &user, time_t time)
    {
        sendUserValue(SET_LASTTIME, user, time);
    }

	void setCeiling(const std::string &user, float ceiling) 
	{
        if (ceiling < -1) THROW_EX(HTCondorValueError, "Ceiling must be greater than -1.");
        sendUserValue(SET_CEILING, user, ceiling);
	}

    void resetUsage(const std::string &user)
    {
        sendUserCmd(RESET_USAGE, user);
    }

    void deleteUser(const std::string &user)
    {
        sendUserCmd(DELETE_USER, user);
    }

    void resetAllUsage()
    {
        Daemon negotiator(DT_NEGOTIATOR, m_addr.c_str());
        bool result;
        {
        condor::ModuleLock ml;
        result = negotiator.sendCommand(RESET_ALL_USAGE, Stream::reli_sock, 0);
        }
        if (!result)
        {
                THROW_EX(HTCondorIOError, "Failed to send RESET_ALL_USAGE command");
        }
    }

    boost::python::list
    getResourceUsage(const std::string &user)
    {
        checkUser(user);

        boost::shared_ptr<Sock> sock = getSocket(GET_RESLIST);
        if (!sock->put(user.c_str()) ||
            !sock->end_of_message())
        {
            sock->close();
            THROW_EX(HTCondorIOError, "Failed to send GET_RESLIST command to negotiator" );
        }
        sock->decode();
        boost::shared_ptr<ClassAdWrapper> ad(new ClassAdWrapper());
        bool result;
        {
        condor::ModuleLock ml;
        result = !getClassAdNoTypes(sock.get(), *ad.get()) || !sock->end_of_message();
        }
        if (result)
        {
            sock->close();
            THROW_EX(HTCondorIOError, "Failed to get classad from negotiator");
        }
        sock->close();

        std::vector<std::string> attrs;
        attrs.push_back("Name");
        attrs.push_back("StartTime");
        return toList(ad, attrs);
    }

    boost::python::list
    getPriorities(bool rollup=false)
    {
        boost::shared_ptr<Sock> sock = getSocket(rollup ? GET_PRIORITY_ROLLUP : GET_PRIORITY);

        sock->decode();
        boost::shared_ptr<ClassAdWrapper> ad(new ClassAdWrapper());
        bool result;
        {
        condor::ModuleLock ml;
        result = !getClassAdNoTypes(sock.get(), *ad.get()) || !sock->end_of_message();
        }
        if (result)
        {
            sock->close();
            THROW_EX(HTCondorIOError, "Failed to get classad from negotiator");
        }
        sock->close();

        std::vector<std::string> attrs;
        attrs.push_back("Name");
        attrs.push_back("Priority");
        attrs.push_back("ResourcesUsed");
        attrs.push_back("Requested");
        attrs.push_back("WeightedResourcesUsed");
        attrs.push_back("PriorityFactor");
        attrs.push_back("BeginUsageTime");
        attrs.push_back("LastUsageTime");
        attrs.push_back("WeightedAccumulatedUsage");
        attrs.push_back("AccountingGroup");
        attrs.push_back("IsAccountingGroup");
        attrs.push_back("AccumulatedUsage");
        return toList(ad, attrs);
    }

private:

    void checkUser(const std::string &user)
    {
        if( user.find('@') == std::string::npos )
        {
            THROW_EX(HTCondorValueError, "You must specify the full name of the submittor you wish (user@uid.domain)");
        }
    }

    boost::shared_ptr<Sock> getSocket(int cmd)
    {
        Daemon negotiator(DT_NEGOTIATOR, m_addr.c_str());
        Sock *raw_sock;
        {
        condor::ModuleLock ml;
        raw_sock = negotiator.startCommand(cmd, Stream::reli_sock, 0);
        }
        boost::shared_ptr<Sock> sock(raw_sock);
        if (!sock.get()) { THROW_EX(HTCondorIOError, "Unable to connect to the negotiator"); }
        return sock;
    }

    void sendUserValue(int cmd, const std::string &user, float val)
    {
        checkUser(user);
        boost::shared_ptr<Sock> sock = getSocket(cmd);

        bool retval;
        {
        condor::ModuleLock ml;
        retval = !sock->put(user.c_str()) || !sock->put(val) || !sock->end_of_message();
        }
        if (retval)
        {
            sock->close();
            THROW_EX(HTCondorIOError, "Failed to send command to negotiator\n" );
        }
        sock->close();
    }

    void sendUserValue(int cmd, const std::string &user, time_t val)
    {
        checkUser(user);
        boost::shared_ptr<Sock> sock = getSocket(cmd);

        bool retval;
        {
        condor::ModuleLock ml;
        retval = !sock->put(user.c_str()) || !sock->put(val) || !sock->end_of_message();
        }
        if (retval)
        {
            sock->close();
            THROW_EX(HTCondorIOError, "Failed to send command to negotiator\n" );
        }
        sock->close();
    }

    void sendUserCmd(int cmd, const std::string &user)
    {
        checkUser(user);
        boost::shared_ptr<Sock> sock = getSocket(cmd);

        bool retval;
        {
        condor::ModuleLock ml;
        retval = !sock->put(user.c_str()) || !sock->end_of_message();
        }
        if (retval)
        {
            sock->close();
            THROW_EX(HTCondorIOError, "Failed to send command to negotiator\n" );
        }
        sock->close();
    }


    std::string m_addr;
    std::string m_version;

};

BOOST_PYTHON_MEMBER_FUNCTION_OVERLOADS(priority_overloads, getPriorities, 0, 1);

void export_negotiator()
{
    class_<Negotiator>("Negotiator",
            R"C0ND0R(
            This class provides a query interface to the *condor_negotiator*.
            It primarily allows one to query and set various parameters in the fair-share accounting.
            )C0ND0R",
        init<boost::python::object>(
            R"C0ND0R(
            :param location_ad: A ClassAd or DaemonLocation describing the *condor_negotiator*
                location and version.  If omitted, the default pool negotiator is assumed.
            :type location_ad: :class:`~classad.ClassAd` or :class:`DaemonLocation`
            )C0ND0R",
            boost::python::args("self", "ad")))
        .def(boost::python::init<>(boost::python::args("self")))
        .add_property("location", &Negotiator::location,
            R"C0ND0R(
            The negotiator to query or send commands to
            :rtype: location :class:`DaemonLocation`
            )C0ND0R")
        .def("setPriority", &Negotiator::setPriority,
            R"C0ND0R(
            Set the real priority of a specified user.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :param float prio: The priority to be set for the user; must be greater-than 0.0.
            )C0ND0R",
            boost::python::args("self", "user", "prio"))
        .def("setFactor", &Negotiator::setFactor,
            R"C0ND0R(
            Set the priority factor of a specified user.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :param float factor: The priority factor to be set for the user; must be greater-than or equal-to 1.0.
            )C0ND0R",
            boost::python::args("self", "user", "factor"))
        .def("setCeiling", &Negotiator::setCeiling,
            R"C0ND0R(
            Set the submitter ceiling of a specified user.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :param float ceiling: The ceiling t be set for the submitter; must be greater-than or equal-to -1.0.
            )C0ND0R",
            boost::python::args("self", "user", "ceiling"))
        .def("setUsage", &Negotiator::setUsage,
            R"C0ND0R(
            Set the accumulated usage of a specified user.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :param float usage: The usage, in hours, to be set for the user.
            )C0ND0R",
            boost::python::args("self", "user", "usage"))
        .def("setBeginUsage", &Negotiator::setBeginUsage,
            R"C0ND0R(
            Manually set the time that a user begins using the pool.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :param int value: The Unix timestamp of initial usage.
            )C0ND0R",
            boost::python::args("self", "user", "value"))
        .def("setLastUsage", &Negotiator::setLastUsage,
            R"C0ND0R(
            Manually set the time that a user last used the pool.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :param int value: The Unix timestamp of last usage.
            )C0ND0R",
            boost::python::args("self", "user", "value"))
        .def("resetUsage", &Negotiator::resetUsage,
            R"C0ND0R(
            Reset all usage accounting of the specified user.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            )C0ND0R",
            boost::python::args("self", "user"))
        .def("deleteUser", &Negotiator::deleteUser,
            R"C0ND0R(
            Delete all records of a user from the Negotiator's fair-share accounting.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            )C0ND0R",
            boost::python::args("self", "user"))
        .def("resetAllUsage", &Negotiator::resetAllUsage,
            R"C0ND0R(
            Reset all usage accounting.  All known user records in the negotiator are deleted.
            )C0ND0R",
            boost::python::args("self"))
        .def("getResourceUsage", &Negotiator::getResourceUsage,
            R"C0ND0R(
            Get the resources (slots) used by a specified user.

            :param str user: A fully-qualified user name (``USER@DOMAIN``).
            :return: List of ads describing the resources (slots) in use.
            :rtype: list[:class:`~classad.ClassAd`]
            )C0ND0R",
            boost::python::args("self", "user"))
        .def("getPriorities", &Negotiator::getPriorities, priority_overloads(
            R"C0ND0R(
            Retrieve the pool accounting information, one per entry.
            Returns a list of accounting ClassAds.

            :param bool rollup: Set to ``True`` if accounting information, as applied to
                hierarchical group quotas, should be summed for groups and subgroups.
            :return: A list of accounting ads, one per entity.
            :rtype: list[:class:`~classad.ClassAd`]
            )C0ND0R",
            boost::python::args("self", "rollup")))
        ;
}

File: bindings/python/htcondor2/_negotiator.py
from typing import List

from ._common_imports import (
    classad,
    Collector,
    DaemonType,
)

from .htcondor2_impl import (
    _negotiator_command,
    _negotiator_command_return,
    _negotiator_command_user,
    _negotiator_command_user_return,
    _negotiator_command_user_value,
    _dprintf_dfulldebug,
)


def _validate_user(user):
    if not isinstance(user, str):
        raise TypeError("user must be a string")
    if '@' not in user:
        # This was HTCondorValueError in version 1.
        raise ValueError("You must specify the submitter (user@uid.domain)")


class Negotiator():
    """
    The negotiator client.  Query and manage resource usage.

    Consult `User Priorities and Negotiation <https://htcondor.readthedocs.io/en/latest/admin-manual/cm-configuration.html#configuration-for-central-managers>`_ before using these functions.
    """

    def __init__(self, location : classad.ClassAd = None):
        """
        :param location:  A ClassAd with a ``MyAddress`` attribute, such as
            might be returned by :meth:`htcondor2.Collector.locate`.  :py:obj:`None` means the
            default pool negotiator.
        """
        if location is None:
            c = Collector()
            location = c.locate(DaemonType.Negotiator)

        if not isinstance(location, classad.ClassAd):
            raise TypeError("location must be a ClassAd")

        self._addr = location['MyAddress']
        # We never actually use this for anything.
        # self._version = location['CondorVersion']


    def deleteUser(self, user : str) -> None:
        """
        Delete all records of an accounting principal from the negotiator’s
        accounting.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        """
        _validate_user(user)
        _negotiator_command_user(self._addr, self._DELETE_USER, user)


    def getPriorities(self, rollup : bool = False) -> List[classad.ClassAd]:
        """
        Retrieve the pool accounting information as a list of
        `accounting ClassAds <http://htcondor.readthedocs.io/en/latest/classad-attributes/accounting-classad-attributes.html>`_.

        :param rollup:  It :py:obj:`True`, the accounting information
            that applies to heirarchical group quotas will be summed for
            groups and subgroups.
        """
        command = self._GET_PRIORITY
        if rollup:
            command = self._GET_PRIORITY_ROLLUP
        returnAd = _negotiator_command_return(self._addr, command)
        return _convert_numbered_attributes_to_list_of_ads(returnAd)


    def getResourceUsage(self, user : str) -> List[classad.ClassAd]:
        """
        Retrieve the resources (slots) assigned to the specificied
        accounting principal as a list of ClassAds.  The names of the
        attributes and their types are currently undocumented.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        """
        _validate_user(user)
        returnAd = _negotiator_command_user_return(self._addr, self._GET_RESLIST, user)
        return _convert_numbered_attributes_to_list_of_ads(returnAd)


    def resetAllUsage(self):
        """
        Set the accumulated usage of all accounting principals to zero.
        """
        _negotiator_command(self._addr, self._RESET_ALL_USAGE)


    def resetUsage(self, user : str) -> None:
        """
        Set the accumulated usage of the specified accounting principal to zero.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        """
        _validate_user(user)
        _negotiator_command_user(self._addr, self._RESET_USAGE, user)


    def setBeginUsage(self, user : str, when : int) -> None:
        """
        Set the beginning of the specified accounting principal's usage.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        :param when:  A Unix timestamp.
        """
        _validate_user(user)
        if isinstance(when, float):
            _dprintf_dfulldebug("setBeginUsage(): `when` specified as float, ignoring fractional part\n")
        _negotiator_command_user_value(self._addr, self._SET_BEGINTIME, user, int(when))


    # This incorrectly -- and brokenly -- takes a float in version 1.
    def setCeiling(self, user : str, ceiling : int) -> None:
        """
        Set the submitter ceiling for the specific accounting principal.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        :param ceiling:  A value greater than or equal to ``-1``.
        """
        _validate_user(user)
        if isinstance(ceiling, float):
            _dprintf_dfulldebug("setCeiling(): `ceiling` specified as float, ignoring fractional part\n")
        if ceiling <= -1:
            # This was HTCondorValueError in version 1.
            raise ValueError("Ceiling must be greater than -1.")
        _negotiator_command_user_value(self._addr, self._SET_CEILING, user, int(ceiling))


    def setLastUsage(self, user : str, when : int) -> None:
        """
        Set the end of the specified accounting principal's usage.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        :param when:  A Unix timestamp.
        """
        _validate_user(user)
        _negotiator_command_user_value(self._addr, self._SET_LASTTIME, user, int(when))


    # This works exactly the same way it did version 1, including the ugly
    # error message from not waiting for the reply.
    def setFactor(self, user : str, factor : float) -> None:
        """
        Set the priority factor of the specified accounting principal.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        :param factor:  A value greater than or equal to ``1.0``.
        """
        _validate_user(user)
        if factor < 1:
            raise ValueError("Priority factors must be >= 1")
        _negotiator_command_user_value(self._addr, self._SET_PRIORITYFACTOR, user, float(factor))


    def setPriority(self, user : str, priority : float) -> None:
        """
        Set the priority of the specified accounting principal.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        :param priority:  A value greater than or equal to ``0.0``.
        """
        _validate_user(user)
        if priority < 0:
            raise ValueError("User priority must be non-negative")
        _negotiator_command_user_value(self._addr, self._SET_PRIORITY, user, float(priority))


    def setUsage(self, user : str, usage : float) -> None:
        """
        Set the usage of the specified accounting principal.

        :param user:  A fully-qualifed (``user@domain``) accounting principal.
        :param usage:  The usage in hours.
        """
        _validate_user(user)
        if usage < 0:
            raise ValueError("Usage must be non-negative")
        _negotiator_command_user_value(self._addr, self._SET_ACCUMUSAGE, user, float(usage))


    _SCHED_VERS          = 400
    _SET_PRIORITY        = _SCHED_VERS + 49
    _GET_PRIORITY        = _SCHED_VERS + 51
    _RESET_USAGE         = _SCHED_VERS + 58
    _SET_PRIORITYFACTOR  = _SCHED_VERS + 59
    _RESET_ALL_USAGE     = _SCHED_VERS + 60
    _GET_RESLIST         = _SCHED_VERS + 63
    _DELETE_USER         = _SCHED_VERS + 82
    _SET_ACCUMUSAGE      = _SCHED_VERS + 94
    _SET_BEGINTIME       = _SCHED_VERS + 95
    _SET_LASTTIME        = _SCHED_VERS + 96
    _GET_PRIORITY_ROLLUP = _SCHED_VERS + 114
    _SET_CEILING         = _SCHED_VERS + 125
    _SET_FLOOR           = _SCHED_VERS + 130


def _convert_numbered_attributes_to_list_of_ads(ad : classad.ClassAd):
    attrs = []
    for attr in ad.keys():
        if attr.endswith("1"):
            attrs.append(attr[:-1])

    l = []
    index = 1
    carry_on = True
    while carry_on:
        carry_on = False

        c = classad.ClassAd()
        for attr in attrs:
            indexedAttr = f"{attr}{index}"
            e = ad.get(indexedAttr)
            if e is not None:
                carry_on = True
                c[attr] = e

        if carry_on:
            l.append(c)

        index += 1

    return l
